CDF Reader

moldover opened the pull request:
Addresses #242 CDF input for the example file.  
Added new unit test which tabulates CDF file.
CVR Source must be set to "CDF" to trigger CDF parsing (otherwise defaults to ES&S).
CDF files must be tabulated individually (too many weird edge cases could arise otherwise).
If CDF is specified candidates are not required to be in config file.
On config load, CDF file is processed to extract candidate names and add them to candidates list.

tarheel commented:
Nice. I'll check this out tomorrow.

tarheel commented on src/main/java/com/rcv/CommonDataFormatReader.java:
```diff
@@ -0,0 +1,162 @@
+                  HashMap selectionPosition = (HashMap) selectionPositionObject;
+                  // and finally the rank
+                  Integer rank = (Integer) selectionPosition.get("Rank");
+                  assert rank != null;
```
and >= 1?

moldover replied:
I guess?  Never 0-index anything!

tarheel commented on src/main/java/com/rcv/CommonDataFormatReader.java:
```diff
@@ -0,0 +1,162 @@
+
+            // create new cast vote record
+            CastVoteRecord newRecord = new CastVoteRecord(computedCastVoteRecordID,
+                ballotID, null, null, rankings);
```
I guess there's no `fullCVRData` equivalent here? Will our existing audit log just not be compatible with CDF input?

moldover replied:
The idea is that CDF _is_ the audit log.

tarheel commented on src/main/java/com/rcv/TabulatorSession.java:
```diff
@@ -188,8 +188,14 @@ private boolean setUpLogging(ContestConfig config) {
+        List<CastVoteRecord> cvrs;
+        if(source.getProvider().equals("CDF")) {
+          CommonDataFormatReader reader = new CommonDataFormatReader(cvrPath);
+          cvrs = reader.parseCVRFile(castVoteRecords);
```
I just noticed that this API is weird. The caller passes in a collection, and then for every CVR that the method parses, it adds it both to the passed-in collection and to a new collection that it returns.

It would be more intuitive if the caller didn't pass in the collection. When the method returns `cvrs`, we can just call `castVoteRecords.addAll(cvrs)` here.

moldover replied:
Yeah, totally agree.  I think it got to the current state incrementally with no real problem, but it should be easy enough to fix.

tarheel commented on src/main/java/com/rcv/ContestConfig.java:
```diff
@@ -84,10 +84,18 @@
+  // function: loadContestConfig
+  // purpose: create ContestConfig from pre-populated rawConfig and default folder
+  // returns: new ContestConfig object if checks pass otherwise null
+  static ContestConfig loadContestConfig(RawContestConfig rawConfig, String sourceDirectory) {
```
Thanks. This is better.

tarheel commented on src/main/java/com/rcv/ContestConfig.java:
```diff
@@ -119,6 +127,7 @@ static ContestConfig loadContestConfig(String configPath) {
         Logger.log(Level.SEVERE, "Failed to create contest config!");
       }
     }
+    config.processCandidateData();
```
Instead of adding this here, you should refactor the lines immediately above here (where you call the constructor) to just call the other `loadContestConfig` instead.

moldover replied:
dig it

tarheel commented on src/main/java/com/rcv/ContestConfig.java:
```diff
@@ -203,6 +213,21 @@ private void validateCvrFileSources() {
+          }
+          if (isTabulateByPrecinctEnabled()) {
+            isValid = false;
+            Logger.log(Level.SEVERE, "Tabulate By Precinct may not be used with CDF files.");
```
nit: tabulateByPrecinct

tarheel commented on src/main/java/com/rcv/ContestConfig.java:
```diff
@@ -203,6 +213,21 @@ private void validateCvrFileSources() {
+            isValid = false;
+            Logger.log(Level.SEVERE, "Tabulate By Precinct may not be used with CDF files.");
+          }
+          continue;
```
Instead of this, why not just put the ES&S checks in an `else` block?

moldover replied:
yeah that's more readable

moldover commented on src/main/java/com/rcv/CommonDataFormatReader.java:
```diff
@@ -0,0 +1,162 @@
+  // function: getCandidates
+  // purpose: returns list of candidates parsed from CDF election json
+  Set<String> getCandidates()
+  {
```
Yeah, I certainly do.  I don't understand why it doesn't work the same for me.

tarheel replied:
It behaves slightly differently for me if I run it manually vs. as a commit hook. But something like this or `if(X)` it should catch either way...

moldover commented on src/main/java/com/rcv/CommonDataFormatReader.java:
```diff
@@ -0,0 +1,162 @@
+    Set<String> candidates = new HashSet<>();
+    try {
+      // try to read in the file
+      HashMap json = JsonParser.readFromFile(filePath, HashMap.class);
```
It could be, but that ends up producing Unckecked assignment warnings.  Without getting deeper into Java generics, this approach removed the warnings and works reasonably.

moldover commented on src/main/java/com/rcv/CommonDataFormatReader.java:
```diff
@@ -0,0 +1,162 @@
+        HashMap election = (HashMap) electionObject;
+        // for each election get the contests it contains
+        ArrayList contestArray = (ArrayList) election.get("Contest");
+        for(Object contestObject : contestArray) {
```
good call

moldover commented on src/main/java/com/rcv/CommonDataFormatReader.java:
```diff
@@ -0,0 +1,162 @@
+import java.util.logging.Level;
+import javafx.util.Pair;
+
+class CommonDataFormatReader {
```
Yeah, I punted on this.  Won't take long but I'd rather get through the additional tests first.

moldover commented:
Just realized you had some more comments I didn't see (thanks Github!).  Will address them later.

tarheel commented:
> Will address them later.

Do you mean that I should wait for you to make them here, or that you'll do them in a future PR?

tarheel commented:
I'm still mystified by GitHub's behavior of seemingly arbitrarily hiding some review comments...

moldover commented on src/main/java/com/rcv/CommonDataFormatReader.java:
```diff
@@ -0,0 +1,162 @@
+
+    List<CastVoteRecord> cvrs = new ArrayList<>();
+    // cvrIndex and fileName are used to generate IDs for cvrs
+    Integer cvrIndex = 0;
```
fixed

moldover commented on src/main/java/com/rcv/CommonDataFormatReader.java:
```diff
@@ -0,0 +1,162 @@
+        // get ballotID
+        String ballotID = (String) CVRObject.get("BallotPrePrintedId");
+        // compute internal ballot ID
+        String computedCastVoteRecordID = String.format("%s(%d)", fileName, ++cvrIndex);
```
Yep

moldover commented on src/main/java/com/rcv/CommonDataFormatReader.java:
```diff
@@ -0,0 +1,162 @@
+        ArrayList CVRSnapshots = (ArrayList) CVRObject.get("CVRSnapshot");
+        for (Object snapshotObject : CVRSnapshots) {
+          HashMap snapshot = (HashMap) snapshotObject;
+          if (snapshot.get("@id").equals(currentSnapshotID)) {
```
This is of course hypothetical, but when I wrote that I was imagining a scenario where a scanner produces an "original" snapshot which has just the raw marks (maybe density or whatever that scanner's thing is) THEN it applies some rules like party-line selection and produces a second snapshot which is "modified" to have all a particular parties selected.  THEN it is handed off to a tabulator.

I will break up that function, it's a bit lengthy and yes, indented :(

moldover commented:
Ok, I've addressed the comments I missed before and this is ready.  I made a new issue for refactoring the CVR reader classes so that is not included.
